donkeycar/parts/pins.py

```diff
@@ -0,0 +1,913 @@
+
+
+def input_pin(pin_provider:str, pin_number:int, pin_scheme:str=PinScheme.BOARD, pull:int=PinPull.PULL_NONE) -> InputPin:
+    """
+    construct an InputPin using the given pin provider
+    """
+    if pin_provider == PinProvider.RPI_GPIO:
+        return InputPinGpio(pin_number, pin_scheme, pull)
+    if pin_provider == PinProvider.PCA9685:
+        raise RuntimeError("PinProvider.PCA9685 does not implement InputPin")
+    if pin_provider == PinProvider.PIGPIO:
+        if pin_scheme != PinScheme.BCM:
+            raise ValueError("Pin scheme must be PinScheme.BCM for PIGPIO")
+        return InputPinPigpio(pin_number, pull)
+    raise RuntimeError("UnknownPinProvider ({})".format(pin_provider))
+
+
+def output_pin(pin_provider:str, pin_number:int, pin_scheme:str=PinScheme.BOARD, i2c_bus:int=0, i2c_address:int=40, frequency_hz:int=60) -> OutputPin:
+    """
+    construct an output pin using the given pin provider
+    """
+    if pin_provider == PinProvider.RPI_GPIO:
+        return OutputPinGpio(pin_number, pin_scheme)
+    if pin_provider == PinProvider.PCA9685:
+        return OutputPinPCA9685(pin_number, pca9685(i2c_bus, i2c_address, frequency_hz))
+    if pin_provider == PinProvider.PIGPIO:
+        if pin_scheme != PinScheme.BCM:
+            raise ValueError("Pin scheme must be PinScheme.BCM for PIGPIO")
+        return OutputPinPigpio(pin_number)
+    raise RuntimeError("UnknownPinProvider ({})".format(pin_provider))
+
+
+def pwm_pin(pin_provider:str, pin_number:int, pin_scheme:str=PinScheme.BOARD, frequency_hz:int=60, i2c_bus:int=0, i2c_address:int=40) -> PwmPin:
+    """
+    construct a PwmPin using the given pin provider
+    """
+    if pin_provider == PinProvider.RPI_GPIO:
+        return PwmPinGpio(pin_number, pin_scheme, frequency_hz)
+    if pin_provider == PinProvider.PCA9685:
+        return PwmPinPCA9685(pin_number, pca9685(i2c_bus, i2c_address, frequency_hz))
+    if pin_provider == PinProvider.PIGPIO:
+        if pin_scheme != PinScheme.BCM:
+            raise ValueError("Pin scheme must be PinScheme.BCM for PIGPIO")
+        return PwmPinPigpio(pin_number, frequency_hz)
+    raise RuntimeError("UnknownPinProvider ({})".format(pin_provider))
+
+
+#
+# ----- RPi.GPIO/Jetson.GPIO implementations -----
+#
+try:
+    import RPi.GPIO as GPIO
+    # lookups to convert abstact api to GPIO values
+    gpio_pin_edge = [None, GPIO.RISING, GPIO.FALLING, GPIO.BOTH]
+    gpio_pin_pull = [None, GPIO.PUD_OFF, GPIO.PUD_DOWN, GPIO.PUD_UP]
+    gpio_pin_scheme = {PinScheme.BOARD: GPIO.BOARD, PinScheme.BCM: GPIO.BCM}
+except ImportError:
+    print("pigpio was not imported.")
+    globals()["GPIO"] = None
+
+
+def gpio_fn(pin_scheme, fn):
+    """
+    Convenience method to call GPIO function
+    using desired pin scheme.  This restores
+    the previous pin scheme, so it is safe to 
+    mix pin schemes.
+    """
+    prev_scheme = GPIO.getmode() or pin_scheme
+    GPIO.setmode(pin_scheme)
```

Comment:
Just an observation here - our PWM actuator parts are multithreaded for performance reasons so they don't clog up the vehicle loop. I hope there is no issue if these parts will talk to the actuators from different threads.

Reply:
Yes, I don't like this pattern where GPIO is a singleton with shared state.  I agree this could be an issue if there are points using different schemes.  One approach would be to require a global scheme to be set then just check against it and runtime fail in pin spec does not match.  Another approach is to use a semaphore/lock to make sure this section is not accessed by more than one thread.